tests/Add-DbaAgDatabase.Tests.ps1

```diff
@@ -38,49 +47,130 @@ Describe "Add-DbaAgDatabase" -Tag "UnitTests" {
     }
 }
 
-Describe "Add-DbaAgDatabase" -Tag "IntegrationTests" {
+Describe "Add-DbaAgDatabase IntegrationTests" -Tag "IntegrationTests" {
     BeforeAll {
-        $null = Get-DbaProcess -SqlInstance $TestConfig.instance3 -Program 'dbatools PowerShell module - dbatools.io' | Stop-DbaProcess -WarningAction SilentlyContinue
-        $server = Connect-DbaInstance -SqlInstance $TestConfig.instance3
-        $agname = "dbatoolsci_addagdb_agroup"
-        $dbname = "dbatoolsci_addagdb_agroupdb"
-        $newdbname = "dbatoolsci_addag_agroupdb_2"
-        $server.Query("create database $dbname")
-        $backup = Get-DbaDatabase -SqlInstance $TestConfig.instance3 -Database $dbname | Backup-DbaDatabase
+        # We want to run all commands in the BeforeAll block with EnableException to ensure that the test fails if the setup fails.
+        $PSDefaultParameterValues['*-Dba*:EnableException'] = $true
+
+        # We remove all the calls to Get-DbaProcess and Stop-DbaProcess as they are not needed with correct tests.
+        #$null = Get-DbaProcess -SqlInstance $TestConfig.instance3 -Program 'dbatools PowerShell module - dbatools.io' | Stop-DbaProcess -WarningAction SilentlyContinue
+
+        # Collect all the created files to be able to remove them ant the end.
+        $filesToRemove = @( )
+
+        # Connect to the instance(s) using Connect-DbaInstance only if needed.
+        # It is saver to use the instance name from $TestConfig to get a fresh new SMO every time.
+        # In the future, we should also test with pre-opened connection.
+        # Use dbatools commands whenever possible, so that these are tested here as well.
```

Comment:
not sure of this one. Our "suite" is pretty much "1-file tests 1-function" so asking for "use dbatools, so if something ELSE crashes it gets spewed out in here" is not going to help pinpointing the cause.
It's true most of our tests are integration tests and not "unittests" but, again, I wouldn't even spell this out for the casual user... if it uses other functions it's good, but also if it uses some "direct queries" (also, usually, they're a bit faster)

Reply:
I see a lot of `$server.Query('...')` which does not throw with a nice error message like `Invoke-DbaQuery` does. More than once it was difficult to get the root cause from the AppVeyor logs. So instead of `$server.Query("create database $dbName")` I would prefer `New-DbaDatabase -SqlInstance $server -Name $dbName`. And sometimes the `$server = Connect...` is just used for one `server.Query("create database ...`. These two lines could just be one.